### DISCUSSION TITLE
Help needed with default resources for graphs

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17863] [Q&A] [answered]
Hi folks!
Looking for advice on what seems pretty basic stuff.

So imagine we have a dagster graph and a dict of pre-defined "default" resources to re-use when constructing jobs from this graph.
Naturally I will just drop resources into the dict, export it if jobs files and use there as intended.
However we are in python world and this will end up with someone editing resources dict in one job without calling `deepcopy(resources)`  and creating pretty classic, but still hard-to-debug issue.

I can think of some solutions for this, like using resources-dict fabrics instead of just dicts, but it still looks like a complication over cool dagster design.
So I wonder is there any built-it mechanism in dagster which deals with "default resources" for graphs and their job-by-job modifications? (yeah, like dependencies injections but for dummies like me)

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1699542481944159?thread_ts=1699542481.944159&cid=C01U954MEER).*

**Comment by jamiedemaria** [comment, marked as answer]
You can set default resources on your `Definitions` object, and then define custom resources on the individual jobs that need them. for example
```python
class Resource1(ConfigurableResource):
    foo: str

@op
def op1(context, my_resource: Resource1):
    context.log.info(my_resource.foo)

@op
def op2(context, my_resource: Resource1):
    context.log.info(my_resource.foo)

@graph
def my_graph():
    op1()
    op2()

job_with_default_resources = my_graph.to_job(name="job_with_default_resources")

job_with_custom_resources = my_graph.to_job(name="job_with_custom_resource", resource_defs={"my_resource": Resource1(foo="hello world")})

defs = Definitions(
    jobs=[job_with_custom_resources, job_with_default_resources],
    resources={
        "my_resource": Resource1(foo="bar")
    }
)
```

With this setup, if you run `job_with_default_resources` the logs will show a print out of `bar` and when you run `job_with_custom_resources` the logs will show a printout of `hello world`
